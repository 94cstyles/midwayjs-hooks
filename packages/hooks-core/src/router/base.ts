import { extname, relative, resolve, sep } from 'upath'

export abstract class AbstractRouter {
  isJavaScriptFile(file: string) {
    const extensions = ['.ts', '.js', '.mjs']
    if (!extensions.includes(extname(file))) {
      return false
    }

    const testExt = [
      '.test.ts',
      '.test.js',
      '.test.mjs',
    ]

    if (testExt.some((ext) => file.endsWith(ext))) {
      return false
    }

    return !file.includes('_client')
  }

  abstract isApiFile(file: string): boolean

  abstract getFunctionId(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string

  abstract functionToHttpPath(
    file: string,
    functionName: string,
    isExportDefault: boolean
  ): string

  protected isPathInside(child: string, parent: string) {
    const relation = relative(parent, child)
    return Boolean(
      relation &&
        relation !== '..' &&
        !relation.startsWith(`..${sep}`) &&
        relation !== resolve(child)
    )
  }
}
