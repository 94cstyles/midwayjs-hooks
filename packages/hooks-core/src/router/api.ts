import { basename, extname, removeExt, toUnix } from 'upath'
import { AbstractRouter } from './base'
import urlJoin from 'proper-url-join'
import some from 'lodash/some'
import { OperatorType } from '../api'
import { API_BASE_PATH } from '../common'
import createJITI from 'jiti'

export type ApiRouterConfig = {
  source?: string
  basePath?: string
}

const jiti = createJITI()

// TODO support manual setup
export class ApiRouter extends AbstractRouter {
  constructor(public config: ApiRouterConfig) {
    super(config.source)
    config.basePath ??= API_BASE_PATH
  }

  isApiFile(file: string): boolean {
    if (
      !super.isJavaScriptFile(file) ||
      !this.isPathInside(toUnix(file), toUnix(this.source))
    ) {
      return false
    }

    try {
      return this.hasExportApiRoutes(jiti(file))
    } catch (error) {
      console.log('require error', error)
      return false
    }
  }

  hasExportApiRoutes(mod: any) {
    return some(
      mod,
      (exp) =>
        typeof exp === 'function' &&
        !!Reflect.getMetadata(OperatorType.Trigger, exp)
    )
  }

  getFunctionId(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string {
    return exportDefault
      ? removeExt(basename(file), extname(file))
      : functionName
  }

  functionToHttpPath(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string {
    return urlJoin(
      this.config.basePath,
      this.getFunctionId(file, functionName, exportDefault),
      { trailingSlash: false }
    )
  }
}
