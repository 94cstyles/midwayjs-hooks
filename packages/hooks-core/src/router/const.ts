export const INDEX = 'index'
export const CATCH_ALL = '/*'

export const DYNAMIC = ':'
