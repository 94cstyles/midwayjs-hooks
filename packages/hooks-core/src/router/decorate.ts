import { basename, extname, removeExt, toUnix } from 'upath'
import { AbstractRouter } from './base'
import urlJoin from 'proper-url-join'
import some from 'lodash/some'
import { OperatorType } from '../decorate/type'
import { DECORATE_BASE_PATH } from '../common'

export type DecoratorRouterConfig = {
  source?: string
  basePath?: string
}

// TODO support manual setup
export class DecorateRouter extends AbstractRouter {
  constructor(public config: DecoratorRouterConfig) {
    super(config.source)
    config.basePath ??= DECORATE_BASE_PATH
  }

  isApiFile(file: string): boolean {
    if (
      !super.isJavaScriptFile(file) ||
      !this.isPathInside(toUnix(file), toUnix(this.source))
    ) {
      return false
    }

    try {
      return this.hasExportApiRoutes(require(file))
    } catch (error) {
      console.log('require error', error)
      return false
    }
  }

  hasExportApiRoutes(mod: any) {
    return some(mod, (exp) => !!Reflect.getMetadata(OperatorType.Trigger, exp))
  }

  getFunctionId(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string {
    return exportDefault
      ? removeExt(basename(file), extname(file))
      : functionName
  }

  functionToHttpPath(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string {
    return urlJoin(
      this.config.basePath,
      this.getFunctionId(file, functionName, exportDefault),
      { trailingSlash: false }
    )
  }
}
