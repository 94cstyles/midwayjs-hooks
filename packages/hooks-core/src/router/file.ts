import inside from 'is-path-inside'
import { kebabCase } from 'lodash'
import { extname, join, relative, removeExt, toUnix } from 'upath'

import { ProjectConfig, Route } from '../types/config'

interface RouterOptions {
  root: string
  projectConfig: ProjectConfig
  useSourceFile: boolean
}

export class FileRouter {
  root: string

  projectConfig: ProjectConfig

  useSourceFile: boolean

  constructor(options: RouterOptions) {
    const { root, projectConfig, useSourceFile } = options
    this.root = root
    this.projectConfig = projectConfig
    this.useSourceFile = useSourceFile
  }

  get source() {
    if (this.useSourceFile) {
      return join(this.root, this.projectConfig.source)
    }
    return join(this.root, this.projectConfig.build.outDir)
  }

  // src/apis/lambda
  getApiDirectory(baseDir: string) {
    return join(this.source, baseDir)
  }

  getRoute(file: string) {
    const { routes } = this.projectConfig
    return routes.find((route) => {
      const dir = this.getApiDirectory(route.baseDir)
      return this.inside(file, dir)
    })
  }

  getGatewayByRoute(route: Route) {
    const gateway = this.projectConfig.gateway.find((adapter) =>
      adapter.is(route)
    )

    if (!gateway) {
      throw new Error(
        `Can't find the correct gateway adapter, please check if midway.config.ts is correct`
      )
    }

    return gateway
  }

  isApiFile(file: string) {
    if (file.endsWith('.test.ts') || file.endsWith('.test.js')) {
      return false
    }

    if (extname(file) !== '.ts' && extname(file) !== '.js') {
      return false
    }

    const route = this.getRoute(file)
    return !!route
  }

  getFunctionId(file: string, functionName: string, isExportDefault: boolean) {
    const relativePath = relative(this.source, file)
    // src/apis/lambda/index.ts -> apis-lambda-index
    const id = kebabCase(removeExt(relativePath, extname(relativePath)))
    const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
    return name.toLowerCase()
  }

  protected inside(child: string, parent: string) {
    return inside(toUnix(child), toUnix(parent))
  }
}
