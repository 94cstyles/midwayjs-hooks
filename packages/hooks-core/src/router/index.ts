export * from './serverless'
export * from './web'
export { getFunctionId, getHTTPMethod } from './util'
