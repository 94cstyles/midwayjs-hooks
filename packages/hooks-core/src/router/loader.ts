import parseFunctionArgs from 'fn-args'
import isFunction from 'lodash/isFunction'
import pickBy from 'lodash/pickBy'
import { join } from 'upath'

import { run } from '@midwayjs/glob'

import {
  PipeFunction,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  FunctionId,
  HooksMiddleware,
  useHooksMiddleware,
} from '../'
import { OperatorProperty } from '../pipe/type'
import { NewFileRouter } from './new-router'

type LoadConfig = {
  root: string
  source: string
}

type BaseTrigger = {
  type: string
  [key: string]: any
}

type Trigger = BaseTrigger & HTTPTriger

interface HTTPTriger extends BaseTrigger {
  type: 'HTTP'
  method:
    | 'get'
    | 'post'
    | 'put'
    | 'delete'
    | 'patch'
    | 'head'
    | 'options'
    | 'all'
  path: string
}

type ApiRoute = {
  function: PipeFunction
  trigger: Trigger
  middleware: HooksMiddleware[]
  functionId: FunctionId
}

export async function loadApiRoutes(config: LoadConfig): Promise<ApiRoute[]> {
  const router = new NewFileRouter({
    root: config.root,
    source: config.source,
    routes: [],
  })

  const files = run([join(router.source, '**/*.{ts,tsx,js,jsx,mjs}')], {
    cwd: router.source,
    ignore: [
      '**/*.test.{ts,tsx,js,jsx,mjs}',
      '**/*.spec.{ts,tsx,js,jsx,mjs}',
      '**/*.d.{ts,tsx}',
      '**/node_modules/**',
    ],
  }).filter((file) => router.isApiFile(file))

  const routes: ApiRoute[] = []
  for (const file of files) {
    const fileRoutes = loadFileApiRoutes(require(file), file, router)
    routes.push(...fileRoutes)
  }
  return routes
}

export function loadFileApiRoutes(
  mod: any,
  file: string,
  router: NewFileRouter
) {
  const apiRoutes: ApiRoute[] = []
  const fileMiddleware = mod?.config?.middleware || []

  const funcs = pickBy<PipeFunction>(mod, isFunction)

  for (let [name, fn] of Object.entries(funcs)) {
    const exportDefault = name === 'default'
    const functionName = exportDefault ? EXPORT_DEFAULT_FUNCTION_ALIAS : name
    const functionId = router.getFunctionId(file, functionName, exportDefault)

    const trigger: Trigger = fn.meta.get(OperatorProperty.Trigger)
    // special case for http trigger
    if (trigger.type === 'HTTP') {
      if (!fn.isPipe) {
        trigger.method = parseFunctionArgs(fn).length > 0 ? 'post' : 'get'
      }
      trigger.path = router.fileToHttpPath(file, functionName, exportDefault)
    }

    const fnMiddleware = fn.meta.get(OperatorProperty.Middleware) || []
    const middleware = fnMiddleware
      .concat(fileMiddleware)
      .map(useHooksMiddleware)

    apiRoutes.push({ function: fn, trigger, functionId, middleware })
  }

  return apiRoutes
}
