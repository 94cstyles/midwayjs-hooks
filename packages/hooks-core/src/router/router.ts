import inside from 'is-path-inside'
import { basename, dirname, extname, join, relative, toUnix } from 'upath'
import chalk from 'chalk'
import { InternalConfig, isProduction } from '..'

const LambdaMethodPrefix = '_'

export class ServerRouter {
  root: string
  routes = new Map<string, string>()

  config: InternalConfig

  constructor(root: string, config: InternalConfig) {
    this.root = root
    this.config = config
  }

  // src/apis
  get source() {
    if (isProduction()) {
      return join(this.root, this.config.build.outDir)
    }
    return join(this.root, this.config.source)
  }

  getRouteConfigBySourceFilePath(sourceFilePath: string) {
    const { routes } = this.config
    const dirs = routes.map((route) => this.getLambdaDirectory(route.baseDir))
    const index = dirs.findIndex((dir) => this.inside(sourceFilePath, dir))
    const route = routes[index]

    if (!route) {
      return null
    }

    return {
      underscore: false,
      ...route,
    }
  }

  // src/apis/lambda
  getLambdaDirectory(baseDir: string) {
    return join(this.source, baseDir)
  }

  isLambdaFile(sourceFilePath: string) {
    if (
      sourceFilePath.endsWith('.test.ts') ||
      sourceFilePath.endsWith('.test.js')
    ) {
      return false
    }

    if (
      extname(sourceFilePath) !== '.ts' &&
      extname(sourceFilePath) !== '.js'
    ) {
      return false
    }

    const route = this.getRouteConfigBySourceFilePath(sourceFilePath)
    return !!route
  }

  protected inside(child: string, parent: string) {
    return inside(toUnix(child), toUnix(parent))
  }

  getBaseUrl(sourceFilePath: string) {
    const url = this.getHTTPPath(sourceFilePath, '', true)
    if (url === '/*') {
      return '/'
    }

    if (url.endsWith('/*')) {
      return url.slice(0, -2)
    }

    return url
  }

  getHTTPPath(
    sourceFilePath: string,
    method: string,
    isExportDefault: boolean
  ) {
    const {
      basePath,
      baseDir,
      underscore,
    } = this.getRouteConfigBySourceFilePath(sourceFilePath)
    const lambdaDirectory = this.getLambdaDirectory(baseDir)

    const { isCatchAllRoutes, filename } = parseFilename(
      basename(sourceFilePath, extname(sourceFilePath))
    )
    const fileRoute = filename === 'index' ? '' : filename
    const methodPrefix = underscore ? LambdaMethodPrefix : ''
    const func = isExportDefault ? '' : `${methodPrefix}${method}`

    const api = toUnix(
      join(
        basePath,
        /**
         * /apis/lambda/index.ts -> ''
         * /apis/lambda/todo/index.ts -> 'todo'
         */
        relative(lambdaDirectory, dirname(sourceFilePath)),
        /**
         * index -> ''
         * demo -> '/demo'
         */
        fileRoute,
        // getTodoList -> _getTodoList
        func,
        isCatchAllRoutes ? '/*' : ''
      )
    )

    /**
     * duplicate routes
     */
    const existPath = this.routes.get(api)
    if (existPath && existPath !== toUnix(sourceFilePath)) {
      duplicateLogger(this.root, existPath, sourceFilePath, api)
    }
    this.routes.set(api, toUnix(sourceFilePath))

    return api
  }
}

function parseFilename(filename: string) {
  const re = /\[\.{3}(.+)]/
  const isCatchAllRoutes = re.test(filename)

  return {
    isCatchAllRoutes,
    filename: isCatchAllRoutes ? re.exec(filename)?.[1] : filename,
  }
}

function duplicateLogger(
  root: string,
  existPath: string,
  currentPath: string,
  api: string
) {
  console.log(
    '[ %s ] Duplicate routes detected. %s and %s both resolve to %s. Reference: %s',
    chalk.yellow('warn'),
    chalk.cyan(relative(root, existPath)),
    chalk.cyan(relative(root, currentPath)),
    chalk.cyan(api),
    'https://www.yuque.com/midwayjs/faas/et7x4k'
  )
}
