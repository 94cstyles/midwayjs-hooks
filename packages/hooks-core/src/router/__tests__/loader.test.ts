import noop from 'lodash/noop'
import { Decorate } from '../../decorate'
import {
  All,
  Del,
  Get,
  Head,
  Options,
  Patch,
  Post,
  Put,
} from '../../decorate/operator/http'
import { Middleware } from '../../decorate/operator/middleware'
import { Operator, OperatorType } from '../../decorate/type'
import { loadFileApiRoutes } from '../loader'
import { FileRouter } from '../router'

const router = new FileRouter({
  root: '/',
  source: 'server',
  routes: [
    {
      baseDir: 'api',
      basePath: '/',
    },
  ],
})

it('load file route with http trigger', () => {
  const routes = loadFileApiRoutes(
    {
      get: Decorate(Get(), async () => {}),
      post: Decorate(Post(), async () => {}),
      put: Decorate(Put(), async () => {}),
      delete: Decorate(Del(), async () => {}),
      patch: Decorate(Patch(), async () => {}),
      head: Decorate(Head(), async () => {}),
      options: Decorate(Options(), async () => {}),
      default: Decorate(All(), async () => {}),
    },
    '/server/api/index.ts',
    router
  )
  expect(routes).toMatchSnapshot()
})

it('load file route with custom trigger', () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      defineMeta({ setProperty: defineProperty }) {
        defineProperty(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
        })
      },
    }
  }

  const routes = loadFileApiRoutes(
    {
      custom: Decorate(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    router
  )

  expect(routes).toMatchSnapshot()
})

it('load middleware', () => {
  const routes = loadFileApiRoutes(
    {
      get: Decorate(Get(), Middleware(noop), async () => {}),
      post: Decorate(Post(), Middleware([noop, noop]), async () => {}),
      config: {
        middleware: [noop],
      },
    },
    '/server/api/index.ts',
    router
  )

  expect(routes[0].middleware.length).toBe(2)
  expect(routes[1].middleware.length).toBe(3)
})
