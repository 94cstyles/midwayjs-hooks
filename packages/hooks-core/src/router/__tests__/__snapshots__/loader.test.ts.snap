// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load file route with http trigger 1`] = `
Array [
  Object {
    "function": [Function],
    "functionId": "api-index-get",
    "middleware": Array [],
    "trigger": Object {
      "method": "GET",
      "path": "/get",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index-post",
    "middleware": Array [],
    "trigger": Object {
      "method": "POST",
      "path": "/post",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index-put",
    "middleware": Array [],
    "trigger": Object {
      "method": "PUT",
      "path": "/put",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index-delete",
    "middleware": Array [],
    "trigger": Object {
      "method": "DELETE",
      "path": "/delete",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index-patch",
    "middleware": Array [],
    "trigger": Object {
      "method": "PATCH",
      "path": "/patch",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index-head",
    "middleware": Array [],
    "trigger": Object {
      "method": "HEAD",
      "path": "/head",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index-options",
    "middleware": Array [],
    "trigger": Object {
      "method": "OPTIONS",
      "path": "/options",
      "type": "HTTP",
    },
  },
  Object {
    "function": [Function],
    "functionId": "api-index",
    "middleware": Array [],
    "trigger": Object {
      "method": "ALL",
      "path": "/",
      "type": "HTTP",
    },
  },
]
`;
